### PR TITLE
Add ref-maybe-const changes to evolution.rst

### PR DESCRIPTION
**Pull request by jabraham17**
Adds a section about ref-maybe-const (also called ref-if-modified) to the language evolution document

[Reviewed by @bmcdonald3]

### CODE REVIEW

File: doc/rst/language/evolution.rst [hunk 1]
```diff
@@ -51,6 +54,96 @@ Additionally, several ``c_string`` methods are deprecated without replacement:
...
+.. _readme-evolution.ref-if-modified-deprecation:
+
+The default intent for arrays and records
+*****************************************
+
+In version 1.32, arrays and records now always have a default intent of
+``const``. Previously, these types would be passed by ``const ref`` intent,
+unless they were modified and then they were passed by ``ref`` intent. This
+change was motivated by improving the consistency across types and making
+potential problems more apparent.
```
Comment (bmcdonald3) [inline doc/rst/language/evolution.rst:66]: I would probably append this to the end of the first sentence: "...of const, meaning that if arrays and records are modified inside of a loop body, they must be specified as `ref` intent via a `with` clause." I see you mention this later, but I think that putting it right at the beginning would be helpful.

Also, I think it would be more clear if we said something like "the compiler would treat these types as `const ref` or `ref`", rather than just "these types were passed".

File: doc/rst/language/evolution.rst [hunk 2]
```diff
@@ -51,6 +54,96 @@ Additionally, several ``c_string`` methods are deprecated without replacement:
...
+   var myInt: int;
+   const myDomain = {1..10};
+   var myArray: [myDomain] int;
+
+   forall i in 2..9 with (ref myInt, ref myArray) {
+     myInt += i;
+     myArray[i] = myArray[i-1] + 1;
+   }
+
+This removes the inconsistency and makes the potential issues much more clear.
```
Comment (bmcdonald3) [inline doc/rst/language/evolution.rst:104]: Nit, but I think I would reword this last part as "calls greater attention to potential race conditions," rather than "makes the potential issues much more clear"

**Review by bmcdonald3 — COMMENTED** [review]
Just a couple of wording comments, but I think those code examples help clarify, thanks for doing this on such short notice Jade!

**Review by bmcdonald3 — APPROVED** [review]
Oops, meant to hit approve.

File: doc/rst/language/evolution.rst (hunk 1 above)
Comment (jabraham17) [reply doc/rst/language/evolution.rst:66]: Rewritten to the following paragraph

```
In version 1.32, arrays and records now always have a default intent of
``const``. This means that if arrays and records are modified inside of a loop
body, a function, or a task body they must use a ``ref`` intent. This also
means that record methods which modify their implicit ``this`` argument must
also use a ``ref`` intent. Previously, the compiler would treat these types as
either ``const ref`` intent or ``ref`` intent, depending on if they were
modified. This change was motivated by improving the consistency across types
and making potential problems more apparent.
```
